kubeflow/trainer/backends/container/backend.py

```diff
@@ -0,0 +1,716 @@
+        runtime is requested in the config.
+
+        Raises RuntimeError if neither Docker nor Podman are available.
+        """
+        if self.cfg.runtime:
+            # User specified a runtime explicitly
+            if self.cfg.runtime == "docker":
+                adapter = DockerClientAdapter(self.cfg.container_host)
+                adapter.ping()
+                logger.info("Using Docker as container runtime")
+                return adapter
+            elif self.cfg.runtime == "podman":
+                adapter = PodmanClientAdapter(self.cfg.container_host)
+                adapter.ping()
+                logger.info("Using Podman as container runtime")
+                return adapter
+        else:
+            # Auto-detect: try Docker first, then Podman
+            try:
+                adapter = DockerClientAdapter(self.cfg.container_host)
+                adapter.ping()
+                logger.info("Using Docker as container runtime")
+                return adapter
+            except Exception as docker_error:
+                logger.debug(f"Docker initialization failed: {docker_error}")
+                try:
+                    adapter = PodmanClientAdapter(self.cfg.container_host)
+                    adapter.ping()
+                    logger.info("Using Podman as container runtime")
+                    return adapter
+                except Exception as podman_error:
+                    logger.debug(f"Podman initialization failed: {podman_error}")
+                    raise RuntimeError(
+                        "Neither Docker nor Podman is available. "
+                        "Please install Docker or Podman, or use LocalProcessBackendConfig instead."
+                    ) from podman_error
+
+    @property
+    def _runtime_type(self) -> str:
+        """Get the runtime type for debugging/logging."""
+        return self._adapter._runtime_type
+
+    # ---- Runtime APIs ----
+    def list_runtimes(self) -> list[types.Runtime]:
+        return list_local_runtimes()
+
+    def get_runtime(self, name: str) -> types.Runtime:
+        return get_local_runtime(name)
+
+    def get_runtime_packages(self, runtime: types.Runtime):
+        """
+        Spawn a short-lived container to report Python version, pip list, and nvidia-smi.
+        """
+        image = self._resolve_image(runtime)
+        self._maybe_pull_image(image)
+
+        command = [
+            "bash",
+            "-lc",
+            "python -c \"import sys; print(f'Python: {sys.version}')\" && "
+            "(pip list || echo 'pip not found') && "
+            "(nvidia-smi || echo 'nvidia-smi not found')",
+        ]
+
+        logs = self._adapter.run_oneoff_container(image=image, command=command)
+        print(logs)
+
+    def train(
+        self,
+        runtime: types.Runtime | None = None,
+        initializer: types.Initializer | None = None,
+        trainer: types.CustomTrainer | types.BuiltinTrainer | None = None,
```

Comment:
Please don't use `|` since we still support Python 3.9 for now. Let's be consistent across backends:
https://github.com/kubeflow/sdk/blob/24f0bbd68b00286e34d8bb8ab84edc9d78480de7/kubeflow/trainer/backends/kubernetes/backend.py#L183